centreon/www/class/centreonMonitoring.class.php

```diff
@@ -145,53 +185,57 @@ public function getServiceStatus($hostList, $objXMLBG, $o, $instance, $hostgroup
-            $rq .= <<<SQL
+        if (! $centreonXMLBGRequest->is_admin) {
+            $accessGroups = $centreonXMLBGRequest->access->getAccessGroups();
+            [$bindValues, $accessGroupsSubQuery] = $this->createMultipleBindQuery(array_keys($accessGroups), ':grp_');
+            $toBind = [...$toBind, ...$bindValues];
+            $query .= <<<SQL
                 
                 INNER JOIN centreon_acl
                     ON centreon_acl.host_id = h.host_id
                     AND centreon_acl.service_id = s.service_id
-                    AND centreon_acl.group_id IN ({$grouplistStr})
+                    AND centreon_acl.group_id IN ({$accessGroupsSubQuery})
                 SQL;
         }
-        $rq .= <<<SQL
+        $query .= <<<SQL
 
             WHERE s.enabled = '1'
                 AND h.enabled = '1'
                 AND h.name NOT LIKE '\_Module\_%'
             SQL;
 
         if ($o === "svcgrid_pb" || $o === "svcOV_pb") {
-            $rq .= " AND s.state != 0 ";
+            $query .= " AND s.state != 0 ";
         } elseif ($o === "svcgrid_ack_0" || $o === "svcOV_ack_0") {
-            $rq .= " AND s.acknowledged = 0 AND s.state != 0 ";
+            $query .= " AND s.acknowledged = 0 AND s.state != 0 ";
         } elseif ($o === "svcgrid_ack_1" || $o === "svcOV_ack_1") {
-            $rq .= " AND s.acknowledged = 1 ";
+            $query .= " AND s.acknowledged = 1 ";
         }
 
-        $rq .= " AND h.name IN (" . $hostList . ") ";
+        $query .= " AND h.host_id IN ({$hostNamesSubQuery}) ";
 
         # Instance filter
-        if ($instance !== -1) {
-            $rq .=  " AND h.instance_id = " . $instance . " ";
+        if ($monitoringServerId !== -1) {
+            $query .=  " AND h.instance_id = :monitoring_server_id";
+            $toBind[':monitoring_server_id'] = $monitoringServerId;
         }
 
-        $rq .= " ORDER BY tri ASC, service_name";
+        $query .= " ORDER BY tri ASC, service_name";
+
+        $serviceDetails = [];
+        $statement = $centreonXMLBGRequest->DBC->prepare($query);
+        $centreonXMLBGRequest->DBC->executePreparedQuery($statement, $toBind);
 
-        $tab = [];
-        $DBRESULT = $objXMLBG->DBC->query($rq);
-        while ($svc = $DBRESULT->fetchRow()) {
-            if (!isset($tab[$svc["name"]])) {
-                $tab[$svc["name"]] = [];
+        while ($result = $statement->fetch(\PDO::FETCH_ASSOC)) {
+            if (! isset($serviceDetails[$result["name"]])) {
+                $serviceDetails[$result["name"]] = [];
             }
-            $tab[$svc["name"]][$svc["service_name"]] = [
-                'state' => $svc["state"],
-                'service_id' => $svc['service_id']
+            $serviceDetails[$result["name"]][$result["service_name"]] = [
+                'state' => $result["state"],
+                'service_id' => $result['service_id']
             ];
         }
-        $DBRESULT->closeCursor();
```

Comment:
You can use closeQuery() of CentreonDB to close the query after the fetch() ;)

Reply:
Yes but for this use case the closeQuery is useless.

Reply:
Why ? it seems to me that closeQuery also frees up memory, no ? I always close the query to be sure to not have errors and to make queries easier to read. But I know it's not obligatory too. So I resolve this thread :)